### ISSUE TITLE
Kubernetes Service Bug

### ISSUE THREAD

**Issue by lengrongfu** [kubernetes/kubernetes#97797]
**What happened**:
1、I started to create a Kubernetes service, the type is NodePort, which has been created successfully, and then I changed the type to ClusterIP, and when I applied again, the error was reported as follows.
```
* spec.ports[0].nodePort: Forbidden: may not be used when `type` is 'ClusterIP'
* spec.ports[1].nodePort: Forbidden: may not be used when `type` is 'ClusterIP'
```

**What you expected to happen**:

After modification, apply again successfully.

**How to reproduce it (as minimally and precisely as possible)**:
1. Successfully create a Service whose type is NodePort.
```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  type: NodePort
  selector:
    app: nginx
  ports:
  - port: 80
    targetPort: 80
```

2. Modify the type to ClusterIP, and an error will be reported at this time.
```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  type: ClusterIP
  selector:
    app: nginx
  ports:
  - port: 80
    targetPort: 80
```
```
The Service "nginx" is invalid: spec.ports[0].nodePort: Forbidden: may not be used when `type` is'ClusterIP'
**Anything else we need to know?**:
```
**Environment**:
- Kubernetes version (use `kubectl version`): 
```shell
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-16T00:04:31Z", GoVersion:"go1.14.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:43:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: `minikube`
- OS (e.g: `cat /etc/os-release`): `mac`
- Kernel (e.g. `uname -a`): `10.15.6 (19G2021)`
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

This bug seems to be relatively simple, I want to try to fix it, can someone guide me? Thanks in advance.

**Comment by Clivern** [comment]
@lengrongfu  it fails on validation here https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L4257.

since k8s assign nodePort value if `spec.type=NodePort`. if you do this 

```
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  type: ClusterIP
  selector:
    app: nginx
  ports:
  - port: 80
    targetPort: 80
    nodePort: 
```

it should work as expected. i think we can consider this as bug or enhancement for the api

**Comment by aojea** [comment]
/sig network

as explained by @Clivern , if you want to mutate the service to ClusterIP, you have to drop the field `nodePort` 

@thockin IIRC there was a similar issue recently, should the apiserver release the `nodePort` when mutating the service?

**Comment by andrewsykim** [comment]
I _think_ this would be addressed in v1.20 with this PR https://github.com/kubernetes/kubernetes/pull/95196 that drops type dependent fields. Haven't tested it though

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/pull/95196/files#diff-2dd9a1ef6872092e519941c8a4870f4c2a71eda096b660510b02afe9b3d19fefR346-R350 in particular fixes this, which is only in 1.20.  It's unlikely that we will want to backport it.
